cpp/FileTransferAgent.cpp

```diff
@@ -398,48 +403,66 @@ RemoteStorageRequestOutcome Snowflake::Client::FileTransferAgent::uploadSingleFi
+  do
   {
-    srcFileStream = m_uploadStream;
-  }
-  else
- {
-    try {
-      fs = ::std::fstream(fileMetadata->srcFileToUpload.c_str(),
-                          ::std::ios_base::in |
-                          ::std::ios_base::binary);
-    }
-    catch(...)
+    if (retry > 0)
     {
-      std::string err= "Could not open source file " + fileMetadata->srcFileToUpload ;
-      throw SnowflakeTransferException(TransferError::FAILED_TO_TRANSFER, err.c_str());
+      unsigned long waitInMicroSecs = (retry << 1) * 100000 ; //First retry 200ms, Second retry 400ms, Third retry 800ms
+      CXX_LOG_DEBUG("Retry count %d, Retrying put file %s.", retry, fileMetadata->srcFileName.c_str());
+      CXX_LOG_DEBUG("Sleeping for %ld micro seconds", waitInMicroSecs );
+#ifdef _WIN32
+      Sleep(waitInMicroSecs/1000);  // sleep time is in milliseconds for windows
+#else
+      usleep(waitInMicroSecs); // sleep time is in microseconds for linux and Mac
+#endif
+    }
+    std::basic_iostream<char> *srcFileStream;
+    ::std::fstream fs;
+
+    if (m_uploadStream) {
+      srcFileStream = m_uploadStream;
+    } else {
+      try {
+        fs = ::std::fstream(fileMetadata->srcFileToUpload.c_str(),
+                            ::std::ios_base::in |
+                            ::std::ios_base::binary);
+      }
+      catch (...) {
+        std::string err = "Could not open source file " + fileMetadata->srcFileToUpload;
+        throw SnowflakeTransferException(TransferError::FAILED_TO_TRANSFER, err.c_str());
+      }
+      srcFileStream = &fs;
     }
-    srcFileStream = &fs;
-  }
 
-  Crypto::CipherIOStream inputEncryptStream(*srcFileStream,
-    Crypto::CryptoOperation::ENCRYPT,
-    fileMetadata->encryptionMetadata.fileKey,
-    fileMetadata->encryptionMetadata.iv,
-    FILE_ENCRYPTION_BLOCK_SIZE);
+    Crypto::CipherIOStream inputEncryptStream(*srcFileStream,
+                                              Crypto::CryptoOperation::ENCRYPT,
+                                              fileMetadata->encryptionMetadata.fileKey,
+                                              fileMetadata->encryptionMetadata.iv,
+                                              FILE_ENCRYPTION_BLOCK_SIZE);
+
+    fileMetadata->recordPutGetTimestamp(FileMetadata::PUT_START);
+    // upload stream
+    outcome = client->upload(fileMetadata, &inputEncryptStream);
+    fileMetadata->recordPutGetTimestamp(FileMetadata::PUT_END);
+    CXX_LOG_DEBUG("File upload done.");
+    if (fs.is_open())
+    {
+      fs.close();
+    }
 
-  fileMetadata->recordPutGetTimestamp(FileMetadata::PUT_START);
-  // upload stream
-  RemoteStorageRequestOutcome outcome = client->upload(fileMetadata,
-                                                       &inputEncryptStream);
-  fileMetadata->recordPutGetTimestamp(FileMetadata::PUT_END);
-  CXX_LOG_DEBUG("File upload done.");
-  if (fs.is_open())
+    m_executionResults->SetTransferOutCome(outcome, resultIndex);
```

Comment:
Is there any way we can log and send to telemetry an http code / Or is that done at a lower level ?

Reply:
libsnowflakeclient does not have inband telemetry. These messages are pushed up to the ODBC and are logged and sent to inband Telemetry from there.